src/Platform/Microsoft.Testing.Extensions.AzureDevOps/AzureDevOpsReporter.cs

```diff
@@ -0,0 +1,177 @@
+
+    /// <inheritdoc />
+    public string Version { get; } = AppVersion.DefaultSemVer;
+
+    /// <inheritdoc />
+    public string DisplayName { get; } = AzureDevOpsResources.DisplayName;
+
+    /// <inheritdoc />
+    public string Description { get; } = AzureDevOpsResources.Description;
+
+    /// <inheritdoc />
+    public Task<bool> IsEnabledAsync()
+    {
+        bool isEnabled = _commandLine.IsOptionSet(AzureDevOpsCommandLineOptions.AzureDevOpsOptionName)
+            && string.Equals(_environment.GetEnvironmentVariable("TF_BUILD"), "true", StringComparison.OrdinalIgnoreCase);
+
+        if (isEnabled)
+        {
+            bool found = _commandLine.TryGetOptionArgumentList(AzureDevOpsCommandLineOptions.AzureDevOpsReportSeverity, out string[]? arguments);
+            if (found && arguments?.Length > 0)
+            {
+                if (string.Equals(arguments[0], "warning", StringComparison.OrdinalIgnoreCase))
+                {
+                    _severity = "warning";
+                }
+            }
+        }
+
+        return Task.FromResult(isEnabled);
+    }
+
+    public async Task ConsumeAsync(IDataProducer dataProducer, IData value, CancellationToken cancellationToken)
+    {
+        if (cancellationToken.IsCancellationRequested)
+        {
+            return;
+        }
+
+        try
+        {
+            if (value is not TestNodeUpdateMessage nodeUpdateMessage)
+            {
+                return;
+            }
+
+            TestNodeStateProperty nodeState = nodeUpdateMessage.TestNode.Properties.Single<TestNodeStateProperty>();
+
+            switch (nodeState)
+            {
+                case FailedTestNodeStateProperty failed:
+                    await WriteExceptionAsync(failed.Explanation, failed.Exception);
+                    break;
+                case ErrorTestNodeStateProperty error:
+                    await WriteExceptionAsync(error.Explanation, error.Exception);
+                    break;
+                case CancelledTestNodeStateProperty cancelled:
+                    await WriteExceptionAsync(cancelled.Explanation, cancelled.Exception);
+                    break;
+                case TimeoutTestNodeStateProperty timeout:
+                    await WriteExceptionAsync(timeout.Explanation, timeout.Exception);
+                    break;
+            }
+
+            return;
+        }
+        catch (OperationCanceledException ex) when (ex.CancellationToken == cancellationToken)
+        {
+            // Do nothing, we're stopping
+        }
+
+        return;
+    }
+
+    private async Task WriteExceptionAsync(string? explanation, Exception? exception)
+    {
+        if (exception == null || exception.StackTrace == null)
+        {
+            return;
+        }
+
+        string message = explanation ?? exception.Message;
+
+        if (message == null)
+        {
+            return;
+        }
+
+        string stackTrace = exception.StackTrace;
```

Comment:
This seems to be relying on MSTest implementation detail where we trim stack traces

Reply:
you mean where we trim Assertion internal stack frames from the stack trace? in that case yup. But I am not sure what to do with that for other frameworks.

Reply:
that sounds expensive, maybe first stack frame that points to a path that starts with the repo root?

Reply:
That could work even if we move the dlls to a separate server, but it feels like we are inventing a way to detect what is user code, and imho the test framework is best suited for that, and all frameworks should provide filtered call stack from the assertions.

Reply:
https://github.com/microsoft/testfx/issues/5272 moving here